### PR TITLE
PWGHF: Add Dplus variables and histograms

### PR DESCRIPTION
**Pull request by fcatalan92**
@vkucera @nzardosh @ginnocen Please have a look. In principle, it is ready to be merged

### CODE REVIEW

File: Analysis/Core/include/Analysis/RecoDecay.h [hunk 1]
```diff
@@ -115,6 +115,19 @@ class RecoDecay
     return res;
   }
 
+  /// FIXME: probably cross and dot products should be in some utility class
+  /// Calculates cross product of vectors.
```
Comment (vkucera) [inline Analysis/Core/include/Analysis/RecoDecay.h]: It should be mentioned that it is implemented for 3-vectors only.

File: Analysis/Core/include/Analysis/RecoDecay.h [hunk 2]
```diff
@@ -388,6 +401,70 @@ class RecoDecay
...
+  /// normalized to its uncertainty
+  /// \param decLenXY decay lenght in {x, y} plane
+  /// \param errDecLenXY error on decay lenght in {x, y} plane
+  /// \param momMother {x, y, z} or {x, y} candidate momentum array
+  /// \param impParProng prong impact parameter
+  /// \param errImpParProng error on prong impact parameter
+  /// \param momProng {x, y, z} or {x, y} prong momentum array
+  /// \return normalized difference between expected and observed impact parameter
+  template <std::size_t N, typename T, typename U, typename V, typename Z>
+  static double NormImpParMeasMinusExpProng(T decLenXY, T errDecLenXY, const array<U, N>& momMother, V impParProng, V errImpParProng, const array<Z, N>& momProng)
```
Comment (vkucera) [inline Analysis/Core/include/Analysis/RecoDecay.h]: The type templates should be more general. I don't see why `decLenXY`, `errDecLenXY` must both be `T`, same for `V` in `impParProng`, `errImpParProng` and `N` in `momMother`, `momProng`.

File: Analysis/Core/include/Analysis/RecoDecay.h [hunk 3]
```diff
@@ -388,6 +401,70 @@ class RecoDecay
...
+  /// \param arrErrImpPar array of errors on prong impact parameter (same order as arrImpPar)
+  /// \param momMom array of {x, y, z} or {x, y} prong momenta (same order as arrImpPar)
+  /// \return maximum normalized difference between expected and observed impact parameters
+  template <std::size_t N, std::size_t M, typename T, typename U, typename V, typename W, typename Z>
+  static double MaxNormImpPar(const T& posPV, const T& posSV, U errDecLenXY, const array<V, M>& momMother, const array<W, N>& arrImpPar, const array<W, N>& arrErrImpPar, const array<array<Z, M>, N>& arrMom)
+  {
+    auto decLenXY = distanceXY(posPV, posSV);
+    double maxNormIP{0.};
+    for (auto iProng = 0; iProng < N; ++iProng) {
+      auto prongNormIP = NormImpParMeasMinusExpProng(decLenXY, (double)errDecLenXY, momMother, arrImpPar[iProng], arrErrImpPar[iProng], arrMom[iProng]);
```
Comment (vkucera) [inline Analysis/Core/include/Analysis/RecoDecay.h]: Why is this needed? `(double)errDecLenXY`

File: Analysis/Core/include/Analysis/RecoDecay.h [hunk 4]
```diff
@@ -388,6 +401,70 @@ class RecoDecay
...
+    auto k = dotProd(flineXY, array{mom[0], mom[1]}) / Pt2(mom);
+    auto dx = flineXY[0] - k * (double)mom[0];
+    auto dy = flineXY[1] - k * (double)mom[1];
+    auto absImpPar = sqrtSumOfSquares(dx, dy);
+    auto fline = array{posSV[0] - posP[0], posSV[1] - posP[1], posSV[2] - posP[2]};
+    auto cross = crossProd(mom, fline);
+    return (cross[2] > 0. ? absImpPar : -1. * absImpPar);
+  }
+
+  /// Calculates the difference between measured and expected track impact parameter
```
Comment (vkucera) [inline Analysis/Core/include/Analysis/RecoDecay.h:425]: Is this an XY calculation (instead of XYZ)? if yes, it should be mentioned in the function name and in the documentation.

File: Analysis/Core/include/Analysis/RecoDecay.h [hunk 5]
```diff
@@ -388,6 +401,70 @@ class RecoDecay
...
+  {
+    // Ported from AliAODRecoDecayHF::Getd0MeasMinusExpProng
+    auto sinThetaP = ((double)momProng[0] * (double)momMother[1] - (double)momProng[1] * (double)momMother[0]) / (Pt(momProng) * Pt(momMother));
+    auto diff = impParProng - (double)decLenXY * sinThetaP;
+    auto errImpParExp = (double)errDecLenXY * sinThetaP;
+    auto errDiff = sqrtSumOfSquares(errImpParProng, errImpParExp);
+    return (errDiff > 0. ? diff / errDiff : 0.);
+  }
+
+  /// Calculates maximum normalized difference between measured and expected impact parameter of candidate prongs
```
Comment (vkucera) [inline Analysis/Core/include/Analysis/RecoDecay.h:447]: Is this an XY calculation (instead of XYZ)? if yes, it should be mentioned in the function name and in the documentation.

File: Analysis/Tasks/PWGHF/taskDPlus.cxx [hunk 6]
```diff
@@ -26,23 +26,28 @@ using namespace o2::aod::hf_cand_prong3;
...
+  OutputObj<TH1F> hdeclengthxy{TH1F("hdeclengthxy", "3-prong candidates;decay length xy (cm);entries", 200, 0., 2.)};
+  OutputObj<TH1F> hnormdlxy{TH1F("hnormdlxy", "3-prong candidates;norm. decay length xy;entries", 80, 0., 80.)};
+  OutputObj<TH1F> hCPA{TH1F("hCPA", "3-prong candidates;cos. pointing angle;entries", 110, -1.1, 1.1)};
+  OutputObj<TH1F> hCPAxy{TH1F("hCPAxy", "3-prong candidates;cos. pointing angle xy;entries", 110, -1.1, 1.1)};
+  OutputObj<TH1F> himpparxy{TH1F("himpparxy", "3-prong candidates;impact par. xy (cm);entries", 200, -1., 1.)};
+  OutputObj<TH1F> hnormip{TH1F("hnormip", "3-prong candidates;norm. IP;entries", 200, -20., 20.)};
+  OutputObj<TH1F> himpparsumsq{TH1F("himpparsumsq", "3-prong candidates;sq. sum of prong imp. par. (cm^2);entries", 100, 0., 1.)};
+  OutputObj<TH1F> hDecLenErr{TH1F("hDecLenErr", "3-prong candidates;decay length error (cm);entries", 100, 0., 1.)};
+  OutputObj<TH1F> hDecLenXYErr{TH1F("hDecLenXYErr", "3-prong candidates;decay length xy error (cm);entries", 100, 0., 1.)};
+  OutputObj<TH1F> hImpParErr{TH1F("hImpParErr", "3-prong candidates;impact parameter error (cm);entries", 100, 0., 1.)};
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskDPlus.cxx]: Why are the histograms reshuffled?
Please fix the names of the new histograms to comply with the conventions.
I will take care of the old ones.

File: Analysis/Tasks/PWGHF/taskDPlus.cxx [hunk 7]
```diff
@@ -54,23 +59,28 @@ struct TaskDPlus {
...
+      hCPA->Fill(candidate.cpa());
+      hCPAxy->Fill(candidate.cpaXY());
+      himpparxy->Fill(candidate.impParXY());
+      hnormip->Fill(candidate.normIP());
+      himpparsumsq->Fill(candidate.impParProngSqSum());
+      hDecLenErr->Fill(candidate.errorDecayLength());
+      hDecLenXYErr->Fill(candidate.errorDecayLengthXY());
+      hImpParErr->Fill(candidate.errorImpactParameter0());
+      hImpParErr->Fill(candidate.errorImpactParameter1());
+      hImpParErr->Fill(candidate.errorImpactParameter2());
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskDPlus.cxx]: Why are the histograms reshuffled?

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Analysis/Core/include/Analysis/RecoDecay.h [hunk 8]
```diff
@@ -388,6 +401,70 @@ class RecoDecay
...
+  /// \param posPV {x, y, z} or {x, y} position of primary vertex
+  /// \param posSV {x, y, z} or {x, y} position of secondary vertex
+  /// \param errDecLenXY error on decay lenght in {x, y} plane
+  /// \param momMother {x, y, z} or {x, y} candidate momentum array
+  /// \param arrImpPar array of prong impact parameters
+  /// \param arrErrImpPar array of errors on prong impact parameter (same order as arrImpPar)
+  /// \param momMom array of {x, y, z} or {x, y} prong momenta (same order as arrImpPar)
+  /// \return maximum normalized difference between expected and observed impact parameters
+  template <std::size_t N, std::size_t M, typename T, typename U, typename V, typename W, typename Z>
+  static double MaxNormImpPar(const T& posPV, const T& posSV, U errDecLenXY, const array<V, M>& momMother, const array<W, N>& arrImpPar, const array<W, N>& arrErrImpPar, const array<array<Z, M>, N>& arrMom)
```
Comment (vkucera) [inline Analysis/Core/include/Analysis/RecoDecay.h]: Same comments as above about the type templates.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Analysis/Tasks/PWGHF/taskDPlus.cxx (hunk 6 above)
Comment (fcatalan92) [reply Analysis/Tasks/PWGHF/taskDPlus.cxx]: This order just makes more sense to me, no specific reason.
For the histogram names, I followed the style already in place. What is the convention?

File: Analysis/Tasks/PWGHF/taskDPlus.cxx (hunk 7 above)
Comment (fcatalan92) [reply Analysis/Tasks/PWGHF/taskDPlus.cxx]: This order just makes more sense to me, no specific reason.

File: Analysis/Core/include/Analysis/RecoDecay.h (hunk 2 above)
Comment (fcatalan92) [reply Analysis/Core/include/Analysis/RecoDecay.h]: I tried to avoid having a number of type templates equal to the number of variables. E. g. `decLenXY` and `errDecLenXY` are related to the same "quantity" so it makes sense that they are of the same type. However, no problem in changing it

File: Analysis/Core/include/Analysis/RecoDecay.h (hunk 3 above)
Comment (fcatalan92) [reply Analysis/Core/include/Analysis/RecoDecay.h]: For the type template strategy that I adopted. Not needed anymore now

File: Analysis/Core/include/Analysis/RecoDecay.h (hunk 4 above)
Comment (fcatalan92) [reply Analysis/Core/include/Analysis/RecoDecay.h:425]: The calculation is based on the `hf_cand::ImpactParameter0` etc. which are defined in the XY plane, so we don't have the 2D and 3D cases as for the cosine of pointing angle. Moreover, we do not specify this for the impact parameters (namely they are not labelled as `ImpactParameterXY...`). What strategy do you suggest?

File: Analysis/Core/include/Analysis/RecoDecay.h (hunk 5 above)
Comment (fcatalan92) [reply Analysis/Core/include/Analysis/RecoDecay.h:447]: See comment above

File: Analysis/Core/include/Analysis/RecoDecay.h (hunk 4 above)
Comment (vkucera) [reply Analysis/Core/include/Analysis/RecoDecay.h:425]: OK, it can stay as it is now.
If at some point there will be a need to use the 3D quantities, we will have to distinguish their names.

File: Analysis/Core/include/Analysis/RecoDecay.h (hunk 5 above)
Comment (vkucera) [reply Analysis/Core/include/Analysis/RecoDecay.h:447]: OK

File: Analysis/Tasks/PWGHF/taskDPlus.cxx (hunk 6 above)
Comment (vkucera) [reply Analysis/Tasks/PWGHF/taskDPlus.cxx]: https://github.com/AliceO2Group/CodingGuidelines

Comment (fcatalan92) [reply Analysis/Tasks/PWGHF/taskDPlus.cxx]: I had already checked the guidelines. Do you mean the [generic naming rules](https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=General_Naming_Rules#General_Naming_Rules)? So, for example, `himpparxy` -> `hImpParXY` (I would keep the abbreviations to avoid very long variables)?

Comment (vkucera) [reply Analysis/Tasks/PWGHF/taskDPlus.cxx]: Yes

**Comment by fcatalan92** [comment]
@vkucera @nzardosh @ginnocen Ready to be merged

File: Analysis/DataModel/include/Analysis/HFSecondaryVertex.h [hunk 9]
```diff
@@ -131,10 +132,12 @@ DECLARE_SOA_DYNAMIC_COLUMN(ImpactParameterProduct, impactParameterProduct, [](fl
...
+DECLARE_SOA_DYNAMIC_COLUMN(ImpactParameterProngSqSum, impactParameterProngSqSum, [](float impParProng0, float impParProng1) { return RecoDecay::sumOfSquares(impParProng0, impParProng1); });
+DECLARE_SOA_DYNAMIC_COLUMN(MaxNormalisedDeltaIP, maxNormalisedDeltaIP, [](float xVtxP, float yVtxP, float xVtxS, float yVtxS, float errDlxy, float pxM, float pyM, float ip0, float errIp0, float ip1, float errIp1, float px0, float py0, float px1, float py1) { return RecoDecay::maxNormalisedDeltaIP(array{xVtxP, yVtxP}, array{xVtxS, yVtxS}, errDlxy, array{pxM, pyM}, array{ip0, ip1}, array{errIp0, errIp1}, array{array{px0, py0}, array{px1, py1}}); });
 // MC matching result:
 // - bit 0: D0(bar) → π± K∓
-DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, uint8_t); // reconstruction level
-DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, uint8_t); // generator level
+DECLARE_SOA_COLUMN(FlagMCMatchRec, flagMCMatchRec, uint8_t); // Rec MC matching result: 0 - not matched, 1 - matched D0(bar)
+DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, uint8_t); // Gen MC matching result: 0 - not matched, 1 - matched D0(bar)
```
Comment (vkucera) [inline Analysis/DataModel/include/Analysis/HFSecondaryVertex.h]: This change should not be there.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Analysis/DataModel/include/Analysis/HFSecondaryVertex.h (hunk 9 above)
Comment (fcatalan92) [reply Analysis/DataModel/include/Analysis/HFSecondaryVertex.h]: It slipped in during the rebase. Fixed

**Comment by jgrosseo** [comment]
@ginnocen I have changed to draft PR because you don't want to have it merged

**Comment by ginnocen** [comment]
hi @fcatalan92, this PR looks fine to me. If there are no stoppers on your side, I would suggest you change the status to ready for review and we approve it.

**Comment by fcatalan92** [comment]
Hi @ginnocen, it is ready to be merged

**Review by ginnocen — APPROVED** [review]

**Comment by ginnocen** [comment]
hi @fcatalan92 perfect thanks. I will merge once the tests have passed. Cheers GM

**Comment by fcatalan92** [comment]
> hi @fcatalan92 perfect thanks. I will merge once the tests have passed. Cheers GM

Thanks, the current errors for `build/AliceO2/O2/o2/macOS` and `build/O2/o2-dataflow` are not related to this PR. 
One is a CCDB test and the other is related to alice3 tracking (@mpuccio)

**Comment by ginnocen** [comment]
hi @fcatalan92. Yes, indeed. I will have to see if I can force the merge even in presence of unrelated errors. If yes, I will do it otherwise Jan will have to do it. GM

**Review by ginnocen — APPROVED** [review]
